In [1]:
#https://towardsdatascience.com/simple-and-multiple-linear-regression-in-python-c928425168f9

import pandas as pd

df = pd.read_csv('/home/alp/Google Drive/Classes/IC230TS/Excel_HW/election-context-2018.csv')

In [2]:
df.head(3)

state     county   fips PctDT PctHC PctOth16 PctMR PctBO PctOth12  \
0  Illinois      Adams  17001  71.3  24.0      4.6  66.7  31.5      1.8   
1  Illinois  Alexander  17003  52.6  44.3      3.1  42.5  56.1      1.4   
2  Illinois       Bond  17005  64.9  27.5      7.6  55.8  41.2      3.0   

  Pct>18yo  ... Unnamed: 30 Unnamed: 31 Unnamed: 32 Unnamed: 33 Unnamed: 34  \
0     76.8  ...         NaN         NaN         NaN         NaN         NaN   
1     75.7  ...         NaN         NaN         NaN         NaN         NaN   
2     78.5  ...         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 35 Unnamed: 36 Unnamed: 37 Unnamed: 38 Unnamed: 39  
0         NaN         NaN         NaN         NaN         NaN  
1         NaN         NaN         NaN         NaN         NaN  
2         NaN         NaN         NaN         NaN         NaN  

[3 rows x 40 columns]

In [3]:
df.tail()

state county fips PctDT PctHC PctOth16 PctMR PctBO PctOth12 Pct>18yo  ...  \
127   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   
128   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   
129   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   
130   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   
131   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   

    Unnamed: 30 Unnamed: 31 Unnamed: 32 Unnamed: 33 Unnamed: 34 Unnamed: 35  \
127         NaN         NaN         NaN         NaN         NaN         NaN   
128         NaN         NaN         NaN         NaN         NaN         NaN   
129         NaN         NaN         NaN         NaN         NaN         NaN   
130         NaN         NaN         NaN         NaN         NaN         NaN   
131         NaN         NaN         NaN         NaN         NaN         NaN   

    Unnamed: 36 Unnamed: 37 Unnamed: 38 Unnamed: 39  
127         NaN         NaN         NaN         NaN  
128         NaN         NaN         NaN         NaN  
129         NaN         NaN         NaN         NaN  
130         NaN         NaN         NaN         NaN  
131         NaN         NaN         NaN         NaN  

[5 rows x 40 columns]

In [4]:
df.columns

Index(['state', 'county', 'fips', 'PctDT', 'PctHC', 'PctOth16', 'PctMR',
       'PctBO', 'PctOth12', 'Pct>18yo', 'PctTurnout', 'white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc', 'Unnamed: 27', 'Unnamed: 28', 'Unnamed: 29',
       'Unnamed: 30', 'Unnamed: 31', 'Unnamed: 32', 'Unnamed: 33',
       'Unnamed: 34', 'Unnamed: 35', 'Unnamed: 36', 'Unnamed: 37',
       'Unnamed: 38', 'Unnamed: 39'],
      dtype='object')

In [5]:
df= df[['state', 'county', 'fips', 'PctDT', 'PctHC', 'PctOth16', 'PctMR',
       'PctBO', 'PctOth12', 'Pct>18yo', 'PctTurnout', 'white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc']]
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 27 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   102 non-null    object 
 1   county                  125 non-null    object 
 2   fips                    126 non-null    object 
 3   PctDT                   125 non-null    object 
 4   PctHC                   124 non-null    object 
 5   PctOth16                123 non-null    object 
 6   PctMR                   122 non-null    object 
 7   PctBO                   121 non-null    object 
 8   PctOth12                120 non-null    object 
 9   Pct>18yo                119 non-null    object 
 10  PctTurnout              118 non-null    object 
 11  white_pct               117 non-null    object 
 12  black_pct               116 non-null    object 
 13  hispanic_pct            115 non-null    object 
 14  nonwhite_pct            114 non-null    ob

Let's set up a regression analysis to see which of these variables is most important in determining pct vote for DT.

In [6]:
df = df.dropna()

In [7]:
df.tail()

state      county   fips PctDT PctHC PctOth16 PctMR PctBO PctOth12  \
97   Illinois   Whiteside  17195  49.3  43.1      7.6  40.3  57.8      1.9   
98   Illinois        Will  17197  44.1  50.5      5.4  46.5  52.0      1.5   
99   Illinois  Williamson  17199  67.7  26.9      5.3  61.5  36.5      2.0   
100  Illinois   Winnebago  17201  47.0  47.1      5.9  46.3  51.9      1.8   
101  Illinois    Woodford  17203  66.5  25.6      7.9  68.7  29.5      1.8   

    Pct>18yo  ... age29andunder_pct age65andolder_pct median_hh_inc  \
97      76.1  ...              36.0              18.9         49151   
98      68.3  ...              41.2              11.2         77507   
99      77.2  ...              36.2              17.5         45902   
100     72.3  ...              38.7              15.6         49468   
101     74.8  ...              38.4              16.3         68040   

    clf_unemploy_pct lesshs_pct lesscollege_pct lesshs_whites_pct  \
97               6.8       11.3            82.7               9.7   
98               6.8        9.2            66.5               4.8   
99               7.6        9.8            77.9               9.6   
100             10.5       12.6            77.6               8.8   
101              5.3        6.0            69.6               5.6   

    lesscollege_whites_pct rural_pct ruralurban_cc  
97                    81.9      37.3           4.0  
98                    64.3       3.9           1.0  
99                    78.0      34.8           3.0  
100                   75.2       7.9           2.0  
101                   69.5      54.9           2.0  

[5 rows x 27 columns]

In [8]:
col_list = df.columns
col_list[3:]

Index(['PctDT', 'PctHC', 'PctOth16', 'PctMR', 'PctBO', 'PctOth12', 'Pct>18yo',
       'PctTurnout', 'white_pct', 'black_pct', 'hispanic_pct', 'nonwhite_pct',
       'foreignborn_pct', 'female_pct', 'age29andunder_pct',
       'age65andolder_pct', 'median_hh_inc', 'clf_unemploy_pct', 'lesshs_pct',
       'lesscollege_pct', 'lesshs_whites_pct', 'lesscollege_whites_pct',
       'rural_pct', 'ruralurban_cc'],
      dtype='object')

In [9]:
for col in col_list[3:]:
    df[col] = df[col].astype(float)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 102 entries, 0 to 101
Data columns (total 27 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   102 non-null    object 
 1   county                  102 non-null    object 
 2   fips                    102 non-null    object 
 3   PctDT                   102 non-null    float64
 4   PctHC                   102 non-null    float64
 5   PctOth16                102 non-null    float64
 6   PctMR                   102 non-null    float64
 7   PctBO                   102 non-null    float64
 8   PctOth12                102 non-null    float64
 9   Pct>18yo                102 non-null    float64
 10  PctTurnout              102 non-null    float64
 11  white_pct               102 non-null    float64
 12  black_pct               102 non-null    float64
 13  hispanic_pct            102 non-null    float64
 14  nonwhite_pct            102 non-null    fl

In [10]:
df.sample(5)

state   county   fips  PctDT  PctHC  PctOth16  PctMR  PctBO  PctOth12  \
65  Illinois   Mercer  17131   56.4   36.0       7.6   45.4   52.8       1.8   
23  Illinois  Edwards  17047   83.0   13.0       4.0   74.5   23.4       2.1   
31  Illinois   Grundy  17063   57.9   34.7       7.4   53.4   44.5       2.1   
59  Illinois  Madison  17119   54.1   38.9       7.0   49.5   48.1       2.4   
43  Illinois  Johnson  17087   76.4   18.8       4.9   69.9   27.7       2.4   

    Pct>18yo  ...  age29andunder_pct  age65andolder_pct  median_hh_inc  \
65      77.9  ...               33.8               20.3        55750.0   
23      76.8  ...               35.2               19.4        47035.0   
31      73.1  ...               39.8               12.9        67162.0   
59      76.7  ...               38.1               15.6        54573.0   
43      81.3  ...               33.9               19.0        44179.0   

    clf_unemploy_pct  lesshs_pct  lesscollege_pct  lesshs_whites_pct  \
65               6.1         8.1             82.5                8.0   
23               7.2        11.0             88.0               10.8   
31               7.5         7.1             78.2                5.9   
59               8.2         7.4             74.3                7.0   
43               8.9        16.6             83.3               11.7   

    lesscollege_whites_pct  rural_pct  ruralurban_cc  
65                    82.6       77.8            2.0  
23                    88.3      100.0            9.0  
31                    78.1       24.4            1.0  
59                    73.6       13.3            1.0  
43                    81.2      100.0            8.0  

[5 rows x 27 columns]

Let's look at a correlation table for the IL county data. . . 

In [11]:
corr1 = df.corr()
corr1.style.background_gradient(cmap='coolwarm',axis=None).format(precision=3)

/tmp/ipykernel_8733/550126231.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  corr1 = df.corr()


In [12]:
import statsmodels.api as sm

In [13]:
y = df['PctDT'] # dependent
x = df['PctMR'] # independent
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctDT   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.799
Method:                 Least Squares   F-statistic:                     401.6
Date:                Wed, 01 Feb 2023   Prob (F-statistic):           8.52e-37
Time:                        11:07:04   Log-Likelihood:                -316.10
No. Observations:                 102   AIC:                             636.2
Df Residuals:                     100   BIC:                             641.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.6683      3.259     -0.512      0.6

PctMR coeff means that for every point support for Romney (2012) increased, there was a 1.13 increase for support of Trump (2016).

So our regression equation is $ PctDT  =  -1.668 + 1.1315 * PctMR $

$ R^2 = 0.801 $ means that this equation explains about 80% of support for Trump based on support for Romney.

How does this compare with support for HC by Obama?

In [14]:
y = df['PctHC']
x = df['PctBO']
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctHC   R-squared:                       0.822
Model:                            OLS   Adj. R-squared:                  0.820
Method:                 Least Squares   F-statistic:                     461.1
Date:                Wed, 01 Feb 2023   Prob (F-statistic):           3.10e-39
Time:                        11:07:04   Log-Likelihood:                -303.71
No. Observations:                 102   AIC:                             611.4
Df Residuals:                     100   BIC:                             616.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -12.5320      2.093     -5.988      0.0

Multivariate regression . . . 

If I take variables that are positively correlated with support for DT, what sort of MLR can we get?

In [15]:
df.columns

Index(['state', 'county', 'fips', 'PctDT', 'PctHC', 'PctOth16', 'PctMR',
       'PctBO', 'PctOth12', 'Pct>18yo', 'PctTurnout', 'white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc'],
      dtype='object')

In [16]:
y = df['PctDT'] # dependent varible
x = df[['white_pct','age65andolder_pct','median_hh_inc','lesshs_pct', 'lesscollege_pct','rural_pct','ruralurban_cc' ]] # independent variable(s)
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)


print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctDT   R-squared:                       0.803
Model:                            OLS   Adj. R-squared:                  0.788
Method:                 Least Squares   F-statistic:                     54.72
Date:                Wed, 01 Feb 2023   Prob (F-statistic):           2.04e-30
Time:                        11:07:04   Log-Likelihood:                -315.50
No. Observations:                 102   AIC:                             647.0
Df Residuals:                      94   BIC:                             668.0
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const               -36.2066     11.42

In [17]:
y = df['PctHC'] # dependent varible
x = df[['black_pct', 'hispanic_pct', 'nonwhite_pct', 'female_pct',
       'age29andunder_pct',  'median_hh_inc','lesshs_pct', 'lesscollege_pct',
       'rural_pct', 'ruralurban_cc']]
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)


print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctHC   R-squared:                       0.838
Model:                            OLS   Adj. R-squared:                  0.820
Method:                 Least Squares   F-statistic:                     47.01
Date:                Wed, 01 Feb 2023   Prob (F-statistic):           1.26e-31
Time:                        11:07:04   Log-Likelihood:                -298.90
No. Observations:                 102   AIC:                             619.8
Df Residuals:                      91   BIC:                             648.7
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                85.5516     25.20

Is IL simlar to the whole nation? Let's look at the national data and see if the regression lines are different.

What happens if we use ALL the demographic variables? Do we increase the R^2?

In [33]:
y = df['PctHC'] # dependent varible
x = df[['white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc']]
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctHC   R-squared:                       0.847
Model:                            OLS   Adj. R-squared:                  0.821
Method:                 Least Squares   F-statistic:                     31.78
Date:                Wed, 01 Feb 2023   Prob (F-statistic):           1.10e-28
Time:                        11:24:11   Log-Likelihood:                -295.87
No. Observations:                 102   AIC:                             623.7
Df Residuals:                      86   BIC:                             665.7
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                      0

So we go from $R^2$ of 83.8 to 86.4. A bit better, but not much. . . 

In [19]:
y = df['PctDT'] # dependent varible
x = df[['Pct>18yo', 'PctTurnout', 'white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc']]
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctDT   R-squared:                       0.855
Model:                            OLS   Adj. R-squared:                  0.825
Method:                 Least Squares   F-statistic:                     29.08
Date:                Wed, 01 Feb 2023   Prob (F-statistic):           4.41e-28
Time:                        11:07:05   Log-Likelihood:                -299.94
No. Observations:                 102   AIC:                             635.9
Df Residuals:                      84   BIC:                             683.1
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                      0

When we do the calculation for DT, we go from $R^2$ of 80.3 to 85.5. This is a bigger gain than for HC, but our first DT MLR didn't have as many variables, so it's hard to compare these.

In [20]:
import pandas as pd
nation = pd.read_excel('/home/alp/Google Drive/Classes/IC230TS/Excel_HW/election-context-2018.xlsx')
nation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3114 entries, 0 to 3113
Data columns (total 43 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3114 non-null   object 
 1   county                  3114 non-null   object 
 2   fips                    3114 non-null   int64  
 3   trump16                 3114 non-null   int64  
 4   pctDT                   3113 non-null   float64
 5   clinton16               3114 non-null   int64  
 6   pctHC                   3113 non-null   float64
 7   otherpres16             3114 non-null   int64  
 8   total16                 3114 non-null   int64  
 9   romney12                3114 non-null   int64  
 10  obama12                 3114 non-null   int64  
 11  otherpres12             3114 non-null   int64  
 12  total12                 3114 non-null   int64  
 13  demsen16                1942 non-null   float64
 14  repsen16                1942 non-null   

In [21]:
nation = nation.fillna(0)

In [22]:
import statsmodels.api as sm

In [23]:
nation.columns

Index(['state', 'county', 'fips', 'trump16', 'pctDT', 'clinton16', 'pctHC',
       'otherpres16', 'total16', 'romney12', 'obama12', 'otherpres12',
       'total12', 'demsen16', 'repsen16', 'othersen16', 'demhouse16',
       'rephouse16', 'otherhouse16', 'demgov16', 'repgov16', 'othergov16',
       'repgov14', 'demgov14', 'othergov14', 'total_population', 'cvap',
       'white_pct', 'black_pct', 'hispanic_pct', 'nonwhite_pct',
       'foreignborn_pct', 'female_pct', 'age29andunder_pct',
       'age65andolder_pct', 'median_hh_inc', 'clf_unemploy_pct', 'lesshs_pct',
       'lesscollege_pct', 'lesshs_whites_pct', 'lesscollege_whites_pct',
       'rural_pct', 'ruralurban_cc'],
      dtype='object')

In [24]:
nation.sample(5)

state      county   fips  trump16      pctDT  clinton16      pctHC  \
2805   Virginia  Buckingham  51029     3950  54.146676       3128  42.878684   
2427  Tennessee    Grainger  47057     6626  82.742258       1154  14.410589   
86     Arkansas       Boone   5009    12235  75.941903       2926  18.161505   
3104    Wyoming    Niobrara  56027     1116  84.931507        115   8.751903   
603    Illinois     Jackson  17077    10843  44.046797      11634  47.260024   

      otherpres16  total16  romney12  ...  age29andunder_pct  \
2805          217     7295      3569  ...          34.051674   
2427          228     8008      5470  ...          33.559812   
86            950    16111     11159  ...          35.650519   
3104           83     1314      1022  ...          36.269015   
603          2140    24617      9864  ...          49.101169   

      age65andolder_pct  median_hh_inc  clf_unemploy_pct  lesshs_pct  \
2805          16.658837        43514.0          8.723637   22.476098   
2427          18.502608        37522.0          8.210721   19.568548   
86            19.728694        38664.0          8.427877   15.021426   
3104          18.254604        40640.0          0.457875   11.054422   
603           13.114145        33845.0          9.022556    7.744271   

      lesscollege_pct  lesshs_whites_pct  lesscollege_whites_pct   rural_pct  \
2805        88.097553          18.501292               87.054264  100.000000   
2427        88.398603          18.775980               88.629533  100.000000   
86          85.032140          15.036225               85.064961   62.198195   
3104        81.009070           7.936508               79.809524  100.000000   
603         64.094744           7.671967               64.256249   37.166628   

      ruralurban_cc  
2805            3.0  
2427            2.0  
86              7.0  
3104            9.0  
603             3.0  

[5 rows x 43 columns]

In [25]:
y = nation['pctDT'] # dependent varible
x = nation[['white_pct','age65andolder_pct','median_hh_inc','lesshs_pct', 'lesscollege_pct','rural_pct','ruralurban_cc' ]] # independent variable(s)
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  pctDT   R-squared:                       0.569
Model:                            OLS   Adj. R-squared:                  0.568
Method:                 Least Squares   F-statistic:                     585.1
Date:                Wed, 01 Feb 2023   Prob (F-statistic):               0.00
Time:                        11:07:06   Log-Likelihood:                -11698.
No. Observations:                3114   AIC:                         2.341e+04
Df Residuals:                    3106   BIC:                         2.346e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const               -43.6468      2.87

In [26]:
y = nation['pctHC'] # dependent varible
x = nation[['black_pct', 'hispanic_pct', 'nonwhite_pct', 'female_pct',
       'age29andunder_pct',  'median_hh_inc','lesshs_pct', 'lesscollege_pct',
       'rural_pct', 'ruralurban_cc']]
x = sm.add_constant(x) 

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  pctHC   R-squared:                       0.626
Model:                            OLS   Adj. R-squared:                  0.625
Method:                 Least Squares   F-statistic:                     518.9
Date:                Wed, 01 Feb 2023   Prob (F-statistic):               0.00
Time:                        11:07:06   Log-Likelihood:                -11380.
No. Observations:                3114   AIC:                         2.278e+04
Df Residuals:                    3103   BIC:                         2.285e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                80.4495      3.91

In [31]:
nation.columns

Index(['state', 'county', 'fips', 'trump16', 'pctDT', 'clinton16', 'pctHC',
       'otherpres16', 'total16', 'romney12', 'obama12', 'otherpres12',
       'total12', 'demsen16', 'repsen16', 'othersen16', 'demhouse16',
       'rephouse16', 'otherhouse16', 'demgov16', 'repgov16', 'othergov16',
       'repgov14', 'demgov14', 'othergov14', 'total_population', 'cvap',
       'white_pct', 'black_pct', 'hispanic_pct', 'nonwhite_pct',
       'foreignborn_pct', 'female_pct', 'age29andunder_pct',
       'age65andolder_pct', 'median_hh_inc', 'clf_unemploy_pct', 'lesshs_pct',
       'lesscollege_pct', 'lesshs_whites_pct', 'lesscollege_whites_pct',
       'rural_pct', 'ruralurban_cc'],
      dtype='object')

In [32]:
y = nation['pctHC'] # dependent varible

x = nation[['white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc']]

x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())



                            OLS Regression Results                            
Dep. Variable:                  pctHC   R-squared:                       0.656
Model:                            OLS   Adj. R-squared:                  0.654
Method:                 Least Squares   F-statistic:                     369.5
Date:                Wed, 01 Feb 2023   Prob (F-statistic):               0.00
Time:                        11:19:13   Log-Likelihood:                -11248.
No. Observations:                3114   AIC:                         2.253e+04
Df Residuals:                    3097   BIC:                         2.263e+04
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     56

In [27]:
# National Data (including IL)
corr1 = nation[['pctDT', 'pctHC','white_pct', 'black_pct', 'hispanic_pct', 'nonwhite_pct',
       'foreignborn_pct', 'female_pct', 'age29andunder_pct',
       'age65andolder_pct', 'median_hh_inc', 'clf_unemploy_pct', 'lesshs_pct',
       'lesscollege_pct', 'lesshs_whites_pct', 'lesscollege_whites_pct',
       'rural_pct', 'ruralurban_cc']].corr()
corr1.style.background_gradient(cmap='coolwarm',axis=None).format(precision=3)

In [28]:
# Illinois Only
corr1 = df[['PctDT', 'PctHC','white_pct', 'black_pct', 'hispanic_pct', 'nonwhite_pct',
       'foreignborn_pct', 'female_pct', 'age29andunder_pct',
       'age65andolder_pct', 'median_hh_inc', 'clf_unemploy_pct', 'lesshs_pct',
       'lesscollege_pct', 'lesshs_whites_pct', 'lesscollege_whites_pct',
       'rural_pct', 'ruralurban_cc']].corr()
corr1.style.background_gradient(cmap='coolwarm',axis=None).format(precision=3)

After typing all the stuff below, I found a simple explaination here: https://medium.com/swlh/interpreting-linear-regression-through-statsmodels-summary-4796d359035a

### Durbin-Watson statistic.
https://www.statsmodels.org/stable/generated/statsmodels.stats.stattools.durbin_watson.html#statsmodels.stats.stattools.durbin_watson
* The null hypothesis of the test is that there is no serial correlation in the residuals.
* The test statistic is approximately equal to 2*(1-r) where r is the sample autocorrelation of the residuals. Thus, for r == 0, indicating no serial correlation, the test statistic equals 2. This statistic will always be between 0 and 4. The closer to 0 the statistic, the more evidence for positive serial correlation. The closer to 4, the more evidence for negative serial correlation.

### Omnibus test for normality. (= _Chi_^2 test)


### kurtosis measures
https://www.simplypsychology.org/kurtosis.html
* Kurtosis is a statistical measure used to describe the degree to which scores cluster in the tails or the peak of a frequency distribution. The peak is the tallest part of the distribution, and the tails are the ends of the distribution.

* There are three types of kurtosis: mesokurtic, leptokurtic, and platykurtic.
 * When kurtosis is equal to 3, the distribution is mesokurtic. This means the kurtosis is the same as the normal distribution, it is mesokurtic (medium peak). The kurtosis of a mesokurtic distribution is neither high nor low, rather it is considered to be a baseline for the two other classifications.
 * Positive excess values of kurtosis (>3) indicate that a distribution is peaked and possess thick tails. Leptokurtic distributions have positive kurtosis values. A leptokurtic distribution has a higher peak (thin bell) and taller (i.e. fatter and heavy) tails than a normal distribution. 
 * Negative excess values of kurtosis (<3) indicate that a distribution is flat and has thin tails.
 * Platykurtic distributions have negative kurtosis values. A platykurtic distribution is flatter (less peaked) when compared with the normal distribution, with fewer values in its shorter (i.e. lighter and thinner) tails.

### Skewness.
https://www.statisticshowto.com/probability-and-statistics/descriptive-statistics/skewness/
* Skewness is a measure of symmetry in a distribution. Actually, it’s more correct to describe it as a measure of lack of symmetry. A standard normal distribution is perfectly symmetrical and has zero skew.
 * A symmetrical distribution has a skew of zero. 
 * A positive result means that your data is positively skewed. (That is, right-tail.)
 * A negative results means that your data is negatively skewed. (Left-tail.)

### Jarque-Bera test of normality.
https://www.statsmodels.org/stable/generated/statsmodels.stats.stattools.jarque_bera.html#statsmodels.stats.stattools.jarque_bera
https://en.wikipedia.org/wiki/Jarque%E2%80%93Bera_test
* The Jarque–Bera test is a goodness-of-fit test of whether sample data have the skewness and kurtosis matching a normal distribution. Samples from a normal distribution have an expected skewness of 0 and an expected excess kurtosis of 0 (which is the same as a kurtosis of 3).
* 
